### PR TITLE
RSDK-3140 - validate depends on feature flag

### PR DESCRIPTION
**Pull request by kim-mishra**
#### Feature flag not set 
```
  "services": [
    {
      "type": "slam",
      "attributes": {
        "delete_processed_data": true,
        "sensors": [
          "rplidar"
        ],
        "use_live_data": true,
        "config_params": {
          "mode": "2d"
        },
        "data_dir": "/Users/kim/workspace/not-repos/slam_3"
      },
      "model": "viam:slam:cartographer",
      "name": "cartographer"
    }
  ],
```

#### Feature flag set to false
```
  "services": [
    {
      "type": "slam",
      "attributes": {
        "delete_processed_data": true,
        "sensors": [
          "rplidar"
        ],
        "use_live_data": true,
        "modularization_v2_enabled": false,
        "config_params": {
          "mode": "2d"
        },
        "data_dir": "/Users/kim/workspace/not-repos/slam_3"
      },
      "model": "viam:slam:cartographer",
      "name": "cartographer"
    }
  ],
```
#### Feature flag set to true
```
  "services": [
    {
      "type": "slam",
      "attributes": {
        "delete_processed_data": true,
        "sensors": [
          "rplidar"
        ],
        "use_live_data": true,
        "modularization_v2_enabled": true,
        "config_params": {
          "mode": "2d"
        },
        "data_dir": "/Users/kim/workspace/not-repos/slam_3"
      },
      "model": "viam:slam:cartographer",
      "name": "cartographer"
    }
  ],
```

#### Feature flag set to true with missing old config params
```
  "services": [
    {
      "type": "slam",
      "attributes": {
        "delete_processed_data": true,
        "sensors": [
          "rplidar"
        ],
        "modularization_v2_enabled": true,
        "config_params": {
          "mode": "2d"
        },
        "data_dir": "/Users/kim/workspace/not-repos/slam_3"
      },
      "model": "viam:slam:cartographer",
      "name": "cartographer"
    }
  ],
```

#### Feature flag missing, with missing old config params (should error)
```
  "services": [
    {
      "type": "slam",
      "attributes": {
        "delete_processed_data": true,
        "sensors": [
          "rplidar"
        ],
        "modularization_v2_enabled": true,
        "config_params": {
          "mode": "2d"
        },
        "data_dir": "/Users/kim/workspace/not-repos/slam_3"
      },
      "model": "viam:slam:cartographer",
      "name": "cartographer"
    }
  ],
```
errors with:
```
2023-05-22T14:02:56.245-0400	ERROR	robot_server	impl/local_robot.go:896	modular config validation error found in component: cartographer	{"error": "rpc error: code = Unknown desc = error validating resource: error validating \"cartographer\": \"use_live_data\" is required"}
```

### CODE REVIEW

File: config/config.go [hunk 1]
```diff
@@ -108,17 +132,19 @@ func GetOptionalParameters(config *Config, defaultPort string,
...
-	useLiveData, err := DetermineUseLiveData(logger, config.UseLiveData, config.Sensors)
-	if err != nil {
-		return "", 0, 0, false, false, false, err
+	useLiveData := false
+	err := errors.New("")
+	if !modularizationV2Enabled {
+		useLiveData, err = DetermineUseLiveData(logger, config.UseLiveData, config.Sensors)
+		if err != nil {
+			return "", 0, 0, false, false, false, err
+		}
```
Comment (kim-mishra) [inline config/config.go:141]: is there a better way to initialize err here. I get an error if I do not declare it before line 137

File: config/config.go [hunk 2]
```diff
@@ -86,8 +100,18 @@ func (config *Config) Validate(path string) ([]string, error) {
...
 	defaultDataRateMsec, defaultMapRateSec int, logger golog.Logger,
 ) (string, int, int, bool, bool, bool, error) {
+	modularizationV2Enabled := false
+	if config.ModularizationV2Enabled == nil {
+		logger.Debug("no modularization_v2_enabled given, continuing with modularization v1")
+	} else {
+		modularizationV2Enabled = *config.ModularizationV2Enabled
+	}
+
+	// do not validate port if mod v2 is enabled
```
Comment (EshaMaharishi) [inline config/config.go:110]: [nit] here and below, comments should be formatted as sentences (capitalization and ending with a period)

File: config/config.go [hunk 3]
```diff
@@ -108,17 +132,19 @@ func GetOptionalParameters(config *Config, defaultPort string,
 		logger.Info("setting slam system to localization mode")
 	}
 
-	useLiveData, err := DetermineUseLiveData(logger, config.UseLiveData, config.Sensors)
-	if err != nil {
-		return "", 0, 0, false, false, false, err
+	var err error
```
[the remark this answers is not in the document]
Comment (kim-mishra) [reply config/config.go:135]: @EshaMaharishi  I couldn't figure out how to get this conditional to work without it. 

This gives me an error that `useLiveData` is not being used, even though it is returned
```
	useLiveData := true
	if !modularizationV2Enabled {
		useLiveData, err := DetermineUseLiveData(logger, config.UseLiveData, config.Sensors)
		if err != nil {
			return "", 0, 0, false, false, false, err
		}
	}
```

and this gives `err is not defined`
```
	useLiveData := true
	if !modularizationV2Enabled {
		useLiveData, err = DetermineUseLiveData(logger, config.UseLiveData, config.Sensors)
		if err != nil {
			return "", 0, 0, false, false, false, err
		}
	}
```

any other ideas?

[the remark this answers is not in the document]
Comment (EshaMaharishi) [reply config/config.go:134]: ah i see, i didn't realize you can't have one of the variables declared already but the other one not: https://stackoverflow.com/a/32628626

ok, it looks good to me as is! feel free to resolve

**Pull request merged**